src/site/content/en/blog/google-search-sw/index.md

```diff
@@ -0,0 +1,463 @@
+
+{% Aside 'note' %}
+While this custom solution is internal to Google, the same general principle can
+be applied to any domain that includes a number of different logical web apps,
+all of whom live under a common URL.
+{% endAside %}
+
+### Personalized results and metrics
+
+Users can sign in to Google Search using their Google Accounts, and their search
+results experience may be customized based on their particular account data.
+Logged in users are identified by specific [browser cookies](https://developer.mozilla.org/en-US/docs/Web/HTTP/Cookies),
+which is a venerable and widely-supported standard.
+
+One downside of using browser cookies, though, is that they are not exposed
+inside of a service worker, and there is no way of automatically examining their
+values and ensuring that they have not changed due to a user logging out or
+switching accounts. (There is effort underway to
+[bring cookie access to service workers](https://developers.google.com/web/updates/2018/09/asynchronous-access-to-http-cookies#welcome_service_workers),
+but as of this writing, the approach is
+[experimental](https://developers.google.com/web/updates/2018/09/asynchronous-access-to-http-cookies#origin-trial)
+and is not widely supported.)
+
+A mismatch between the service worker's view of the current logged in user and
+the actual user logged in to the Google Search web interface could lead to
+incorrectly personalized search results or misattributed metrics and logging.
+Any of those failure scenarios would be a serious issue for the Google Search
+team.
+
+**Solution**
+
+Rather than wait for experimental APIs to launch and provide direct access to
+the browser's cookies inside of a service worker, the Google Search team went
+with a stop-gap solution: whenever a page controlled by the service worker is
+loaded, the page will read the relevant cookies and use
+[`postMessage()`](https://developer.mozilla.org/en-US/docs/Web/API/Worker/postMessage)
+to send them to the service worker.
+
+The service worker can then check the current cookie value against the value
+that it expects, and if there's a mismatch, it will take steps to purge any
+user-specific data from its storage, and reload the search results page without
+any incorrect personalization.
+
+The specific steps that the service worker takes to reset things to a baseline
+are particular to Google Search's requirements, but the same general approach
+may be useful to other developers who deal with personalized data keyed off of
+browsers cookies.
+
+### Experiments and dynamism
+
+As mentioned, the Google Search team relies heavily on running experiments in
+production, and testing the effects of new code and features in the real world
+before turning them on by default. This can be a bit of a challenge with a
+static service worker that relies heavily on cached data, since opting users in
+and out of experiments often requires communication with the backend server.
+
+**Solution**
```

Comment:
Suggestion to retitle solution headers to also be level three, and to call out the solution:

### Solution: dynamically generated service worker script